api/server/user/data.go

```diff
@@ -176,63 +176,172 @@ func (s *Service) CidInfo(ctx context.Context, req *userPb.CidInfoRequest) (*use
-			LatestPushedStorageConfig: rpcConfig,
-		}
+
+	type source struct {
+		cid            string
+		currentStorage *ffs.StorageInfo
+		queuedJobs     []ffs.StorageJob
+		executingJob   *ffs.StorageJob
+	}
+
+	var sources []source
+
+	for cid := range storageConfigs {
+		d := source{cid: cid.String()}
+
 		info, err := i.Show(cid)
 		if err != nil && err != api.ErrNotFound {
 			return nil, status.Errorf(codes.Internal, "getting storage info: %v", err)
 		} else if err == nil {
-			cidInfo.CurrentStorageInfo = toRPCStorageInfo(info)
-		}
-		queuedJobs := i.QueuedStorageJobs(cid)
-		rpcQueudJobs := make([]*userPb.StorageJob, len(queuedJobs))
-		for i, job := range queuedJobs {
-			rpcJob, err := toRPCJob(job)
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
-			}
-			rpcQueudJobs[i] = rpcJob
+			d.currentStorage = &info
 		}
-		cidInfo.QueuedStorageJobs = rpcQueudJobs
+
+		d.queuedJobs = i.QueuedStorageJobs(cid)
+
 		executingJobs := i.ExecutingStorageJobs(cid)
 		if len(executingJobs) == 1 {
 			// There is exactly one job in the slice because we specified a cid
 			// and there can be only one executing job per cid at a time.
-			rpcJob, err := toRPCJob(executingJobs[0])
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
-			}
-			cidInfo.ExecutingStorageJob = rpcJob
-		} else {
+			d.executingJob = &executingJobs[0]
+		} else if len(executingJobs) > 1 {
 			log.Warnf("received %d executing jobs when there should be 1", len(executingJobs))
 		}
-		finalJobs := i.LatestFinalStorageJobs(cid)
-		if len(finalJobs) > 0 {
-			rpcJob, err := toRPCJob(finalJobs[0])
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
-			}
-			cidInfo.LatestFinalStorageJob = rpcJob
+
+		sources = append(sources, d)
+	}
+
+	extractJobTime := func(d source) int64 {
+		var time int64
+		if d.executingJob != nil {
+			time = d.executingJob.CreatedAt
 		}
-		successfulJobs := i.LatestSuccessfulStorageJobs(cid)
-		if len(successfulJobs) > 0 {
-			rpcJob, err := toRPCJob(successfulJobs[0])
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
+		for _, job := range d.queuedJobs {
+			if job.CreatedAt > time {
+				time = job.CreatedAt
 			}
-			cidInfo.LatestSuccessfulStorageJob = rpcJob
 		}
-		res = append(res, cidInfo)
+		return time
+	}
+
+	extractCurrentStorageTime := func(d source) int64 {
+		var time int64
+		if d.currentStorage != nil {
+			return d.currentStorage.Created.Unix()
+		}
+		return time
 	}
-	sort.Slice(res, func(a, b int) bool {
-		return res[a].Cid < res[b].Cid
+
+	sort.Slice(sources, func(a, b int) bool {
+		jobTimeA := extractJobTime(sources[a])
+		jobTimeB := extractJobTime(sources[b])
+		currentStorageTimeA := extractCurrentStorageTime(sources[a])
+		currentStorageTimeB := extractCurrentStorageTime(sources[b])
```

Comment:
This could be moved below the two ifs to avoid calculating if any of the ifs match.